### PR TITLE
feat: throw error for invalid interpolations at build-time to avoid mistakes

### PR DESCRIPTION
**Pull request by satya164**
image

image

### CODE REVIEW

File: src/__tests__/__snapshots__/preval.test.js.snap [hunk 1]
```diff
@@ -368,3 +368,39 @@ exports[`throws codeframe error when evaluation fails 1`] = `
...
+[0m [90m 7 | [39m[32m\`[39m[33m;[39m[0m"
+`;
+
+exports[`throws when interpolation evaluates to undefined 1`] = `
+"<<DIRNAME>>/source.js: The expression evaluated to 'undefined', which is probably a mistake. If you want it to be inserted into CSS, explicitly cast or transform the value to a string, e.g. - 'String(fontSize)'.
+[0m [90m 4 | [39m[0m
+[0m [90m 5 | [39m[36mconst[39m [33mTitle[39m [33m=[39m styled[33m.[39mh1[32m\`[39m[0m
+[0m[31m[1m>[22m[39m[90m 6 | [39m[32m  font-size: \${fontSize};[39m[0m
+[0m [90m   | [39m               [31m[1m^[22m[39m[0m
+[0m [90m 7 | [39m[32m\`[39m[33m;[39m[0m"
```
Comment (thymikee) [inline src/__tests__/__snapshots__/preval.test.js.snap]: these snapshots would use `strip-ansi` for readability

File: src/react/styled.js [hunk 2]
```diff
@@ -58,10 +58,31 @@ function styled(tag: React.ComponentType<*> | string) {
...
 
         Object.keys(vars).forEach(name => {
-          const [value, unit = ''] = vars[name];
-          style[`--${name}`] = `${
-            typeof value === 'function' ? value(props) : value
-          }${unit}`;
+          const [result, unit = ''] = vars[name];
+          const value = typeof result === 'function' ? result(props) : result;
+
+          if (process.env.NODE_ENV !== 'production') {
```
Comment (thymikee) [inline src/react/styled.js]: extract it to a helper invariant

File: src/babel/extract.js [hunk 3]
```diff
@@ -344,12 +344,36 @@ module.exports = function extract(babel: any, options: Options) {
                   : { line: end.line, column: end.column + 1 },
               };
 
+              // If value is null or undefined, it's probably a mistake
+              const throwIfInvalid = (value: any) => {
```
Comment (thymikee) [inline src/babel/extract.js]: maybe move this to the top scope and pass `ex` as an argument to avoid unnecessary recreation of this function

**Comment by satya164** [comment]
> But isn't it a breaking change

Yeah, but the published version is a beta so...

> I'd think about making it a deprecation warning

It's very easy not to notice warnings amongst the webpack output logs though

**Comment by thymikee** [comment]
You're right about beta :)

**Pull request merged**